crates/ruff_python_semantic/src/model.rs

```diff
@@ -257,6 +257,27 @@ impl<'a> SemanticModel<'a> {
             .is_some_and(|binding| binding.kind.is_builtin())
     }
 
+    /// Return `true` if `member` is a reference to `builtins.$target`,
+    /// i.e. either `object` (where `object` is not overridden in the global scope),
+    /// or `builtins.object` (where `builtins` is imported as a module at the top level)
+    pub fn match_builtin_expr(&self, expr: &Expr, symbol: &str) -> bool {
+        debug_assert!(!symbol.contains('.'));
+        if self.seen_module(Modules::BUILTINS) {
```

Comment:
This is the case where I don't like our clippy rule that switches the `if` and` else if the if condition is negated. Naturally, I would expect that the "fast" path comes first (and that's also what branch predictors assume by default, as far as I know). 

You could try to use 

```
        // fast path
        if !self.seen_module(Modules::BUILTINS) {
            return matches!(expr, Expr::Name(ast::ExprName { id, .. }) if id == symbol && self.is_builtin(symbol));
        }

        self.resolve_qualified_name(expr)
            .is_some_and(|qualified_name| match qualified_name.segments() {
                ["builtins" | "", member] => *member == symbol,
                _ => false,
            })
```

which gets you around the clippy rule and lays out the code so that the "default" branch comes first.

Reply:
Heh, that's what I had originally! But @charliermarsh complained ;) https://github.com/astral-sh/ruff/pull/10919#discussion_r1564197567

Reply:
I misread that version which suggests to me that others will do the same in the future. I find this much clearer, but don't feel strongly and not important enough to debate it so defer to you. If you use that version, I'd suggest at least adding comments and a newline between the closing `}` and the method call as above.

Reply:
To be clear, I'm personally fine with either; I can see both points of view here!

Reply:
Me too :)

Reply:
I'll leave this as-is for now; the status quo wins ;)